examples/phase_estimator.py

```diff
@@ -0,0 +1,151 @@
+RMS Error: 0.0177
+
+Estimation with 8qubits.
+Actual, Estimation (Raw binary)
+0.0000, 0.0000 (00000000)
+0.1000, 0.1016 (00011010)
+0.2000, 0.1992 (00110011)
+0.3000, 0.3008 (01001101)
+0.4000, 0.3984 (01100110)
+0.5000, 0.5000 (10000000)
+0.6000, 0.6016 (10011010)
+0.7000, 0.6992 (10110011)
+0.8000, 0.8008 (11001101)
+0.9000, 0.8984 (11100110)
+RMS Error: 0.0011
+"""
+
+
+import numpy as np
+import cirq
+
+
+def phase_op(phi):
+    """An example unitary 1-qubit operation U with an eigen vector |0>
+    and an eigen value exp(2*Pi*i*phi)
+    """
+
+    gate = cirq.SingleQubitMatrixGate(
+        matrix=np.array([[np.exp(2*np.pi*1.0j*phi), 0], [0, 1]]))
+    return gate
+
+
+class QftInverse(cirq.Gate, cirq.CompositeGate):
+    """Quantum gate for the inverse Quantum Fourier Transformation
+    """
+
+    def default_decompose(self, qubits):
+        """A quantum circuit with the following structure.
+
+        ---H----R_2---R_3---R_4--------------------------------------
+                 |     |     |
+        ---------@-----|-----|-----H----R_2---R_3--------------------
+                       |     |           |     |                       = QFT_inv
+        ---------------@-----|-----------@-----|-----H----R_2--------
+                             |                 |           |
+        ---------------------@-----------------@-----------@-----H---
+
+        where Controlled-R_k is equivalent to CZ**(-1/2^(k-1))
+        (and also to Rot11Gate(rads=-2Pi/2^k)).
+        The number of qubits can be arbitrary.
+        """
+
+        qubits = list(qubits)
+        while len(qubits) > 0:
+            q_head = qubits.pop(0)
+            yield cirq.H(q_head)
+            for i, qubit in enumerate(qubits):
+                yield (cirq.CZ**(-1/2.0**(i+1)))(qubit, q_head)
+
+
+def run_estimate(phi, qnum, repeats):
+    """Construct the following phase estimator circuit and execute simulations.
+
+                                     ---------
+    ---H---------------------@------|         |---M--- [m4]:lowest bit
+                             |      |         |
+    ---H---------------@-----|------|         |---M--- [m3]
+                       |     |      | QFT_inv |
+    ---H---------@-----|-----|------|         |---M--- [m2]
+                 |     |     |      |         |
+    ---H---@-----|-----|-----|------|         |---M--- [m1]:highest bit
+           |     |     |     |       ---------
+    -------U----U^2---U^4---U^8--------------------------
+
+    The measurement results M=[m1, m2,...] are translated to the estimated
+    phase with the following formula:
+    phi = m1*(1/2) + m2*(1/2)^2 + m3*(1/2)^3 + ...
+    """
+
+    qubits = [None] * qnum
+    for i in range(len(qubits)):
+        qubits[i] = cirq.GridQubit(0, i)
+    ancilla = cirq.GridQubit(0, len(qubits))
+
+    circuit = cirq.Circuit.from_ops(
+        cirq.H.on_each(qubits),
+        [cirq.ControlledGate(phase_op((2**i)*phi))(qubits[qnum-i-1], ancilla)
```

Comment:
I think it takes away from the example to explicitly say "phi" in the circuit creation part. Instead, you should pass in an "unknown operation" (which is actually a MatrixGate or what have you, as long as it supports exponentiation) and then in this code you would do `cirq.ControlledGate(unknown_gate**(2**i)).on(...)`.

Reply:
ok, so I will make `run_estimate` to receive `unknown_gate` instead of `phi`. And for clarification, I will move the definition of example gate `phase_op` to the part of `experiment`.